shared/src/ledger/storage/traits.rs

```diff
@@ -159,6 +165,51 @@ impl<'a, H: StorageHasher + Default> SubTreeWrite for &'a mut Amt<H> {
     }
 }
 
+impl<'a> SubTreeRead for &'a BridgePoolTree {
+    fn subtree_has_key(&self, key: &Key) -> Result<bool, Error> {
+        self.contains_key(key)
+            .map_err(|err| Error::MerkleTree(err.to_string()))
+    }
+
+    fn subtree_membership_proof(
+        &self,
+        keys: &[Key],
+        values: Vec<MerkleValue>,
+    ) -> Result<MembershipProof, Error> {
+        let values = values
+            .into_iter()
+            .filter_map(|val| match val {
+                MerkleValue::BridgePoolTransfer(transfer) => Some(transfer),
+                _ => None,
+            })
+            .collect();
+        self.get_membership_proof(keys, values)
+            .map(Into::into)
+            .map_err(|err| Error::MerkleTree(err.to_string()))
+    }
+}
+
+impl<'a> SubTreeWrite for &'a mut BridgePoolTree {
+    fn subtree_update(
+        &mut self,
+        key: &Key,
+        value: MerkleValue,
+    ) -> Result<Hash, Error> {
+        if let MerkleValue::BridgePoolTransfer(_) = value {
+            self.insert_key(key)
+                .map_err(|err| Error::MerkleTree(err.to_string()))
+        } else {
+            Err(Error::InvalidValue)
+        }
+    }
```

Comment:
We should revisit these traits in some other PR, `key` is deriveable from `value` so we shouldn't need to pass in both, at least for `BridgePoolTree`. I'm guessing for other impls this isn't the case and the key and value aren't related.

Reply:
I've changed this method to ignore the keys.